src/base/opt/exterr.rs

```diff
@@ -364,6 +339,65 @@ impl<'a, Target: Composer> OptBuilder<'a, Target> {
     }
 }
 
+
+//------------ LossyOctets ---------------------------------------------------
+
+/// A octets wrapper that displays its content as a lossy UTF-8 sequence.
+#[derive(Clone)]
+struct LossyOctets<Octs>(Octs);
+
+impl<Octs: AsRef<[u8]>> LossyOctets<Octs> {
+    fn as_slice(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl<Octs> From<Octs> for LossyOctets<Octs> {
+    fn from(src: Octs) -> Self {
+        Self(src)
+    }
+}
+
+impl<Octs> AsRef<Octs> for LossyOctets<Octs> {
+    fn as_ref(&self) -> &Octs {
+        &self.0
+    }
+}
+
+impl<Octs: AsRef<[u8]>> fmt::Display for LossyOctets<Octs> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut source = self.0.as_ref();
+        loop {
+            match str::from_utf8(source) {
+                Ok(s) => {
+                    f.write_str(s)?;
+                    break;
+                }
+                Err(err) => {
+                    let (good, mut tail) = source.split_at(err.valid_up_to());
+                    f.write_str(
+                        // Safety: valid UTF8 for this part was confirmed
+                        // above.
+                        unsafe {
+                            str::from_utf8_unchecked(good)
+                        }
+                    )?;
+                    f.write_str("\u{fffd}")?;
```

Comment:
While not caused by this PR, I really would have appreciated a tiny comment to say "Output the unicode <?> replacement character" as I had to Google unicode fffd...

Reply:
Oh, sorry, missed this comment over the big green “approved“ line.